design/core_collection_operations.md

```diff
@@ -0,0 +1,175 @@
+# Design of core collection operations
+
+This document describes an inital design of the core, general data collections operations - e.g.
+operations that aren't specific to a particular collection.
+
+## Assumptions
+
+* Collection information can change at any time, and previous versions are not necessarily
+  accessible.
+  * They should be accessible, at least, to database admins, and possibly via admin API methods.
+* Collections can theoretically be reverted to prior versions at any time.
+* We expect < 1000 collections over the lifetime of the collections infrastructure.
+* There may be many versions of collections as they are updated and adjusted over time.
+  * Including adding new data products to collections.
+* Data product data structures / UIs / etc for collections will usually, but not always,
+  be applicable to other collections as well.
+
+## Collection operations / versions
+
+* Save a collection version.
+  * Administration only operation.
+  * Once saved, a collection version is immutable and cannot be overwritten.
+    * Trying to save a collection with an existing version will result in an error.
+  * A collection version *is not* immediately active.
+    * At some point we may wish to supply admin endpoints for accessing alternate collection
+      versions.
+  * The user specifies the version, which to the collections API is an opaque string. Any
+    semantics are supplied by the user.
+    * Note that this means semantic versions will not sort in order - the DB and API would need
+      specific schemas / code to do that.
+* Activate a collection version.
+  * Administration only operation.
+  * Can activate any version at any time.
+  * The active version is the only version accessible to users.
+  * Separating the saving and activation steps means we can save a collection document,
+    load data products for that collection, and then activate the collection when the data
+    products are ready.
+  * It also means we can trivially roll back changes.
+* Get a collection version.
+  * Public operation.
+  * Initially, this is only the active collection.
+    * At some point we may wish to supply admin endpoints for accessing alternate collection
+      versions.
```

Comment:
should we just build this in from the beginning? It seems like a light lift now, but it could maybe become a heavier lift if we do it as a refactor later on

Reply:
No, since it's not required for the first demo UI. I'd expect this to happen sooner rather than later though. I'm not concerned about it being a heavy lift